src/transport/ExchangeMgr.h

```diff
@@ -0,0 +1,276 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the classes corresponding to CHIP Exchange management.
+ *
+ */
+
+#pragma once
+
+#include <support/BitFlags.h>
+#include <support/DLLUtil.h>
+#include <system/SystemTimer.h>
+#include <transport/SecureSessionMgr.h>
+
+#define EXCHANGE_CONTEXT_ID(x) ((x) + 1)
+
+using namespace chip::Transport;
+
+namespace chip {
+
+using System::PacketBuffer;
+
+class ExchangeManager;
+
+enum class ExFlagValues : uint16_t
+{
+    kFlagInitiator        = 0x0001, /// This context is the initiator of the exchange.
+    kFlagConnectionClosed = 0x0002, /// This context was associated with a Connection.
+    kFlagResponseExpected = 0x0004, /// If a response is expected for a message that is being sent.
+    kFlagMsgRcvdFromPeer =
+        0x0008, /// When set, signifies that at least one message has been received from peer on this exchange context.
+};
+
+/**
+ *  @class ExchangeContext
+ *
+ *  @brief
+ *    This class represents an ongoing conversation (ExchangeContext) between two or more nodes.
+ *    It defines methods for encoding and communicating CHIP messages within an ExchangeContext
+ *    over various transport mechanisms, for example, TCP, UDP, or CHIP Reliable Messaging.
+ *
+ */
+class DLL_EXPORT ExchangeContext
+{
+    friend class ExchangeManager;
+
+public:
+    typedef uint32_t Timeout; /**< Type used to express the timeout in this ExchangeContext, in milliseconds */
+
+    ExchangeManager * ExchangeMgr; /**< [READ ONLY] Owning exchange manager. */
+    uint64_t PeerNodeId;           /**< [READ ONLY] Node ID of peer node. */
+    uint16_t ExchangeId;           /**< [READ ONLY] Assigned exchange ID. */
+    void * AppState;               /**< Pointer to application-specific state object. */
+    bool AllowDuplicateMsgs;       /**< Boolean indicator of whether duplicate messages are allowed for a given exchange. */
+    uint32_t RetransInterval;      /**< Time between retransmissions (in milliseconds); 0 disables retransmissions. */
+    Timeout ResponseTimeout;       /**< Maximum time to wait for response (in milliseconds); 0 disables response timeout. */
+
+    enum
+    {
+        kSendFlag_AutoRetrans    = 0x0001, /**< Used to indicate that automatic retransmission is enabled. */
```

Comment:
Can we add more documentation regarding the purpose of these flags? The current information for some of the flags do not tell enough to understand the use of it.

Reply:
Added documentation to all public APIs

Reply:
This is reserved for future use, I removed the extra flags for future use to avoid confusion.